File: ruby_peer2/identity.rb
# identity.rb
require 'openssl'
require 'json'
require 'base64'
require 'socket'
require_relative 'cryptography'

IDENTITY_PATH = "identity.json"
KEY_PATH = "ecdsa_key.pem"

class PeerIdentity
  attr_reader :username, :key

  def initialize
    @username = nil
    @key = nil
  end

  def setup
    load_or_create_key
    load_or_create_username
  end

  def create_identity
    create_key
    create_username
    puts "\n👤 Your identity information (share with peers):"
    puts "Username: #{@username}"
    puts "Public Key:\n#{public_key_pem}"
  end

  def create_key
    @key = Cryptography.generate_key
    File.write(KEY_PATH, @key.to_pem)
    puts "🔐 New ECDSA key pair generated and saved."
  end

  def create_username
    print "Enter your username: "
    @username = gets.chomp.strip
    File.write(IDENTITY_PATH, { username: @username }.to_json)
    puts "👤 Username '#{@username}' saved."
  end

  def load_or_create_key
    if File.exist?(KEY_PATH)
      @key = OpenSSL::PKey::EC.new(File.read(KEY_PATH))
    else
      create_key
    end
  end

  def load_or_create_username
    if File.exist?(IDENTITY_PATH)
      data = JSON.parse(File.read(IDENTITY_PATH))
      @username = data["username"]
    else
      create_username
    end
  end

  def public_key_pem
    Cryptography.public_key_to_pem(@key)
  end

  def sign_username
    Cryptography.sign(@key, @username)
  end

  def identity_payload
    {
      username: @username,
      public_key: public_key_pem,
      signature: Base64.strict_encode64(sign_username)
    }.to_json
  end

  def rotate_key
    unless File.exist?(IDENTITY_PATH) && File.exist?(KEY_PATH)
      puts "❌ Identity or private key not found."
      return nil
    end

    # Load current identity info
    identity = JSON.parse(File.read(IDENTITY_PATH))
    username = identity["username"]

    # ✅ Load old private key BEFORE overwriting it
    old_key = OpenSSL::PKey::EC.new(File.read(KEY_PATH))

    # 🔐 Generate new ECDSA key pair
    new_key = Cryptography.generate_key
    new_pubkey_pem = Cryptography.public_key_to_pem(new_key)

    # ✅ Normalize the new public key PEM to avoid newline issues
    new_pub_pem = new_pubkey_pem.gsub("\r\n", "\n")

    puts "✍️ new_pub_pem being signed:"
    puts new_pub_pem.inspect
    puts "🔑 SHA256:", OpenSSL::Digest::SHA256.hexdigest(new_pub_pem)

    # ✅ Sign the normalized PEM with old private key using SHA256
    signature = Cryptography.sign(old_key, new_pub_pem)

    # ✅ Now save the new private key and updated identity
    File.write(KEY_PATH, new_key.to_pem)
    File.write(IDENTITY_PATH, JSON.pretty_generate({
      "username" => username,
      "public_key" => new_pub_pem
    }))

    puts "🔁 Key rotation complete. New public key stored."

    {
      "username" => username,
      "new_key" => new_pub_pem,
      "signature" => Base64.strict_encode64(signature)
    }
  end

  # Send authentication payload to peer after ECDH
  def send_authentication(peer_ip, peer_port, session_key)
    signature = Cryptography.sign(@key, session_key)

    payload = {
      username: @username,
      public_key: public_key_pem,
      signature: Base64.strict_encode64(signature)
    }.to_json

    socket = TCPSocket.new(peer_ip, peer_port)
    socket.write("A")
    socket.write([payload.bytesize].pack("N"))
    socket.write(payload)

    response = socket.read(1)
    socket.close

    return response == "A"
  rescue => e
    puts "❌ Send of authentication package failed: #{e.message}"
    return false
  end
end

File: python_peer2/identity.py
import os
import json
from pathlib import Path
from discover import discover_peers
from cryptography.hazmat.primitives import serialization
import socket
import base64
from encryption_utils import generate_key, public_key_to_pem, sign, verify

BASE_DIR = Path(__file__).parent.resolve()
KEY_PATH = BASE_DIR / "ecdsa_key.pem"
IDENTITY_PATH = BASE_DIR / "identity.json"

def identity_exists():
    """Check if both identity file and key file exist"""
    identity_exists = IDENTITY_PATH.exists()
    key_exists = KEY_PATH.exists()
    
    if identity_exists and key_exists:
        return True
    elif identity_exists or key_exists:
        # Partial/broken identity state
        print("⚠️ WARNING: Identity files in inconsistent state.")
        print(f"- Identity file: {'✅ Exists' if identity_exists else '❌ Missing'}")
        print(f"- Key file: {'✅ Exists' if key_exists else '❌ Missing'}")
        return False
    else:
        return False

def cleanup_identity():
    """Remove any existing identity files"""
    try:
        if IDENTITY_PATH.exists():
            IDENTITY_PATH.unlink()
            print(f"✅ Removed {IDENTITY_PATH.name}")
        
        if KEY_PATH.exists():
            KEY_PATH.unlink()
            print(f"✅ Removed {KEY_PATH.name}")
            
        return True
    except Exception as e:
        print(f"❌ Error cleaning up identity files: {e}")
        return False

def ensure_identity_exists():
    """Check if identity exists, and create one if it doesn't"""
    if not identity_exists():
        print("❗ No identity found or identity files are incomplete.")
        
        # Clean up any partial identity files first
        if IDENTITY_PATH.exists() or KEY_PATH.exists():
            cleanup_choice = input("Clean up partial identity files? (y/n): ").strip().lower()
            if cleanup_choice == 'y':
                if not cleanup_identity():
                    print("❌ Failed to clean up identity files. Cannot proceed.")
                    return False
            else:
                print("❌ Cannot proceed with partial identity files.")
                return False
                
        create_choice = input("Create a new identity? (y/n): ").strip().lower()
        if create_choice == 'y':
            create_identity()
            # Verify that identity was created successfully
            if identity_exists():
                return True
            else:
                print("❌ Failed to create identity.")
                return False
        else:
            print("❌ Cannot proceed without an identity.")
            return False
    return True

def create_identity():
    # Check if identity already exists
    identity_file_exists = IDENTITY_PATH.exists()
    key_file_exists = KEY_PATH.exists()
    
    if identity_file_exists or key_file_exists:
        print("⚠️ Some identity files already exist.")
        print(f"- Identity file: {'✅ Exists' if identity_file_exists else '❌ Missing'}")
        print(f"- Key file: {'✅ Exists' if key_file_exists else '❌ Missing'}")
        
        cleanup_choice = input("Clean up existing files and create new identity? (y/n): ").strip().lower()
        if cleanup_choice == 'y':
            if cleanup_identity():
                print("✅ Previous identity cleaned up. Creating new identity...")
            else:
                print("❌ Failed to clean up existing identity files.")
                return False
        else:
            print("❌ Cannot create new identity without cleaning up existing files.")
            return False

    try:
        # Generate EC private key using our utility function
        key = generate_key()
        with open(KEY_PATH, "wb") as f:
            f.write(key.private_bytes(
                encoding=serialization.Encoding.PEM,
                format=serialization.PrivateFormat.TraditionalOpenSSL,
                encryption_algorithm=serialization.NoEncryption()
            ))
        print(f"🔐 New ECDSA key saved to {KEY_PATH.name}")

        # Ask for username
        username = input("Enter your username: ").strip()

        # Extract public key in PEM format using our utility function
        pub_pem = public_key_to_pem(key)

        # Save identity information
        with open(IDENTITY_PATH, "w") as f:
            json.dump({
                "username": username,
                "public_key": pub_pem
            }, f)
        print(f"👤 Identity saved to {IDENTITY_PATH.name}")

        # Print identity info for sharing
        print("\n👤 Your identity information (share with peers):")
        print(f"Username: {username}")
        print("Public Key:\n" + pub_pem)
        
        return True
    except Exception as e:
        print(f"❌ Error creating identity: {e}")
        cleanup_identity()  # Clean up any partial files created
        return False

def sign_session_key(session_key: bytes):
    # Load private key
    with open(KEY_PATH, "rb") as f:
        private_key = serialization.load_pem_private_key(f.read(), password=None)

    # Sign the session key using our utility function
    signature = sign(private_key, session_key)
    print("🔐 Signing session key...")
    print("🔑 Session Key (hex):", session_key.hex())
    print("✍️ Signature (hex):", signature.hex())

    # Load identity
    with open(IDENTITY_PATH, "r") as f:
        identity = json.load(f)

    # Print public key and username
    print("👤 Username:", identity["username"])
    print("📤 Public Key (PEM):\n" + identity["public_key"])

    return {
        "username": identity["username"],
        "public_key": identity["public_key"],
        "signature": signature.hex()
    }
def send_identity_to_ruby(ip, port, identity_payload):
    print("📤 Sending identity authentication payload to Ruby peer...")

    try:
        with socket.create_connection((ip, port), timeout=5) as sock:
            message = {
                "username": identity_payload["username"],
                "public_key": identity_payload["public_key"],
                "signature": identity_payload["signature"]
            }

            msg_str = json.dumps(message)

            # ✅ First, send the 'A' command byte
            sock.send(b"A")

            # ✅ Then send the JSON payload
            sock.sendall(msg_str.encode())

            print("✅ Authentication payload sent.")
            response = sock.recv(4096).decode().strip()
            return response == "A"

    except Exception as e:
        print("❌ Failed to send identity:", e)



def rotate_public_key():
    if not IDENTITY_PATH.exists() or not KEY_PATH.exists():
        print("❌ Cannot rotate key: identity or private key not found.")
        return None

    # Load current identity
    with open(IDENTITY_PATH, "r") as f:
        identity = json.load(f)
    username = identity["username"]
    old_pubkey = identity["public_key"]

    # Load old private key
    with open(KEY_PATH, "rb") as f:
        old_private_key = serialization.load_pem_private_key(f.read(), password=None)

    # Generate new key pair using our utility function
    new_private_key = generate_key()
    # Get PEM format using our utility
    new_pubkey_pem = public_key_to_pem(new_private_key)

    # Sign the new public key with old private key using our utility
    signature = sign(old_private_key, new_pubkey_pem)

    # Save new private key to file
    with open(KEY_PATH, "wb") as f:
        f.write(new_private_key.private_bytes(
            encoding=serialization.Encoding.PEM,
            format=serialization.PrivateFormat.TraditionalOpenSSL,
            encryption_algorithm=serialization.NoEncryption()
        ))

    # Save updated identity
    with open(IDENTITY_PATH, "w") as f:
        json.dump({
            "username": username,
            "public_key": new_pubkey_pem
        }, f, indent=2)

    # Build migration message
    migrate_msg = {
        "username": username,
        "new_key": new_pubkey_pem,
        "signature": base64.b64encode(signature).decode()
    }

    print("🔁 Identity key rotated.")
    print("📤 Send this migration message to your peers:")
    print(json.dumps(migrate_msg, indent=2))

    return migrate_msg  # So you can send it automatically if you want

def notify_peers_of_rotation(migrate_msg):
    peers = discover_peers()
    if not peers:
        print("❌ No peers found to notify.")
        return

    print("\nChoose peer(s) to notify:")
    for i, peer in enumerate(peers, 1):
        print(f"{i}. {peer['name']} @ {peer['ip']}:{peer['port']}")
    selected = input("Enter peer numbers separated by commas (or 'a' for all): ").strip()

    if selected.lower() == 'a':
        selected_peers = peers
    else:
        try:
            indexes = [int(i.strip()) - 1 for i in selected.split(',')]
            selected_peers = [peers[i] for i in indexes if 0 <= i < len(peers)]
        except:
            print("❌ Invalid selection.")
            return

    for peer in selected_peers:
        try:
            with socket.create_connection((peer["ip"], peer["port"]), timeout=5) as sock:
                msg_bytes = json.dumps(migrate_msg).encode()
                sock.send(b"M")
                sock.send(len(msg_bytes).to_bytes(4, 'big'))
                sock.sendall(msg_bytes)
                response = sock.recv(1)
                if response == b"M":
                    print(f"✅ {peer['name']} accepted your new key.")
                else:
                    print(f"⚠️ {peer['name']} rejected your key migration.")
        except Exception as e:
            print(f"❌ Failed to notify {peer['name']}: {e}")




if __name__ == "__main__":
    create_identity()
